build_plugin.py

```diff
@@ -0,0 +1,182 @@
+        if settings.update_file in {'.user.js', '.meta.js'}:
+            append_line('updateURL', path + settings.update_file)
+        append_line('downloadURL', path + '.user.js')
+    append_line('id', name)
+    append_line('namespace', settings.namespace)
+    if keys.isdisjoint({'match', 'include'}):
+        append_line('match', settings.match)
+    append_line('grant', 'none')
+    meta.append('// ==/UserScript==\n')
+    return '\n'.join(meta), is_plugin
+
+
+def readfile(filename):
+    return io.open(filename, 'r', encoding='utf-8-sig').read()
+
+
+def multi_line(text):
+    return ('\n' + text).replace('\\', r'\\').replace('\n', '\\\n').replace("'", r"\'") #"'
+
+
+def load_image(filename):
+    mtype = 'svg+xml' if os.path.splitext(filename) == '.svg' else 'png'
+    return 'data:image/%s;base64,' % mtype + (
+        base64.b64encode(io.open(filename, 'rb').read()).decode('utf8')
+    )
+
+
+def imgrepl(match):
+    return load_image(match.group('filename'))
+
+
+def wrap_iife(filename):
+    name, _ = os.path.splitext(os.path.basename(filename))
+    return """
+// *** module: %s ***
+(function () {
+var log = ulog('%s');
+%s
+
+})();
+""" % (filename, name, readfile(filename))
+
+
+def bundle_code(_):
+    return '\n\n;\n\n'.join(map(wrap_iife, sorted(glob.glob('code/*.js'))))
+
+
+quote = "'%s'"
+
+def expand_template(match):
+    kw, filename = match.groups()
+    if kw == 'include_raw':
+        return """// *** included: %s ***
+%s
+
+""" % (filename, readfile(filename))
+    elif kw == 'include_string':
+        return quote % multi_line(readfile(filename))
+    elif kw == 'include_img':
+        return quote % load_image(filename)
+    elif kw == 'include_css':
+        pattern = r'(?<=url\()["\']?(?P<filename>[^)#]+?)["\']?(?=\))'
+        css = re.sub(pattern, imgrepl, readfile(filename))
+        return quote % multi_line(css)
+    return quote % getattr(settings, kw)
+
+
+def split_filename(path):
+    filename = os.path.basename(path)
+    return filename.split('.')[0]
+
+
+def process_file(source, out_dir, dist_path=None):
+    """Generate .user.js/.meta.js pair from given source file.
+
+    Resulted file(s) put into out_dir (if specified, otherwise - use current).
+    dist_path component is for add to @downloadURL/@updateURL.
+    """
+    try:
+        meta, script = readfile(source).split('\n\n', maxsplit=1)
+    except ValueError:
+        raise Exception('%s: wrong input: empty line expected after metablock' % source)
+    name = split_filename(source)
+    meta, is_plugin = fill_meta(meta, name, dist_path)
+    settings.plugin_id = name
+
+    filename = os.path.join(out_dir, name) if out_dir else name
+    if settings.update_file == '.meta.js':
+        io.open(filename + settings.update_file, 'w', encoding='utf8').write(meta)
+
+    script = re.sub(r"'@bundle_code@';", bundle_code, script)
+    wrapper = import_module(settings.plugin_wrapper)
+    template = r"'@(\w+)(?::([0-9a-zA-Z_./-]+))?@'"
```

Comment:
Ooh. Uh...)
At least add a comment, please, that it's for finding `@type:string@` tags